In [ ]:
# If you haven't installed cognee or nest_asyncio yet, uncomment and run the lines below:
# %pip install cognee
# %pip install nest_asyncio

In [1]:
import os
import glob
import nest_asyncio

import cognee
from cognee.api.v1.search import SearchType
from cognee.api.v1.visualize.visualize import visualize_graph

nest_asyncio.apply()

# Prerequisites:
# 1. Copy `.env.template` and rename it to `.env`.
# 2. Add your OpenAI API key to the `.env` file in the `LLM_API_KEY` field:
#    LLM_API_KEY = "your_key_here"


In [2]:
# Clears all data and system state

await cognee.prune.prune_data()     
await cognee.prune.prune_system(metadata=True)  

INFO:cognee.infrastructure.databases.graph.networkx.adapter:Graph deleted successfully./Users/handekafkas/Documents/local-code/new-cognee/cognee/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
INFO:cognee.infrastructure.databases.relational.sqlalchemy.SqlAlchemyAdapter:Database deleted successfully.

In [3]:
pdf_dir = os.path.join(os.getcwd(), "data", "IT-MANAGER")

pdf_files = glob.glob(os.path.join(pdf_dir, "*.pdf"))[:10]

if not pdf_files:
    print(f"No PDF files found in {pdf_dir}")
else:
    print(f"Found {len(pdf_files)} PDF file(s) to process:")
    for pdf_file in pdf_files:
        print(f"- {os.path.basename(pdf_file)}")

Found 4 PDF file(s) to process:
- 10641230.pdf
- 11584809.pdf
- 10553553.pdf
- 13405733.pdf


In [4]:
if not pdf_files:
    print("No PDFs to add. Please check your file paths.")
else:
    print("Adding PDFs to cognee...")
    for pdf_file in pdf_files:
        print(f"Processing {os.path.basename(pdf_file)}...")
        await cognee.add(pdf_file)

print("All specified PDF files have been added.")


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.WARNING:langfuse:Langfuse client is disabled since no public_key was provided as a parameter or environment variable 'LANGFUSE_PUBLIC_KEY'. See our docs: https://langfuse.com/docs/sdk/python/low-level-sdk#initialize-client13:22:35 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Adding PDFs to cognee...
Processing 10641230.pdf...


INFO:run_tasks(tasks: [Task], data):Pipeline run started: `4b84e400-23fc-5976-bbb4-f8ee303eed81`INFO:run_tasks(tasks: [Task], data):Coroutine task started: `resolve_data_directories`INFO:run_tasks(tasks: [Task], data):Coroutine task started: `ingest_data`

User 0780f453-73e0-4a3f-9440-7ae826795a51 has registered.


/Users/handekafkas/Documents/local-code/new-cognee/cognee/.venv/lib/python3.11/site-packages/dlt/destinations/impl/sqlalchemy/merge_job.py:194: SAWarning: Table 'file_metadata' already exists within the given MetaData - not copying.
  staging_table_obj = table_obj.to_metadata(
/Users/handekafkas/Documents/local-code/new-cognee/cognee/.venv/lib/python3.11/site-packages/dlt/destinations/impl/sqlalchemy/merge_job.py:229: SAWarning: implicitly coercing SELECT object to scalar subquery; please use the .scalar_subquery() method to produce a scalar subquery.
  order_by=order_dir_func(order_by_col),
INFO:run_tasks(tasks: [Task], data):Coroutine task completed: `ingest_data`INFO:run_tasks(tasks: [Task], data):Coroutine task completed: `resolve_data_directories`INFO:run_tasks(tasks: [Task], data):Pipeline run completed: `4b84e400-23fc-5976-bbb4-f8ee303eed81`INFO:run_tasks(tasks: [Task], data):Pipeline run started: `4b84e400-23fc-5976-bbb4-f8ee303eed81`INFO:run_tasks(tasks: [Task], data):Coroutin

Processing 11584809.pdf...
Processing 10553553.pdf...


INFO:run_tasks(tasks: [Task], data):Coroutine task completed: `ingest_data`INFO:run_tasks(tasks: [Task], data):Coroutine task completed: `resolve_data_directories`INFO:run_tasks(tasks: [Task], data):Pipeline run completed: `4b84e400-23fc-5976-bbb4-f8ee303eed81`INFO:run_tasks(tasks: [Task], data):Pipeline run started: `4b84e400-23fc-5976-bbb4-f8ee303eed81`INFO:run_tasks(tasks: [Task], data):Coroutine task started: `resolve_data_directories`INFO:run_tasks(tasks: [Task], data):Coroutine task started: `ingest_data`INFO:run_tasks(tasks: [Task], data):Coroutine task completed: `ingest_data`INFO:run_tasks(tasks: [Task], data):Coroutine task completed: `resolve_data_directories`INFO:run_tasks(tasks: [Task], data):Pipeline run completed: `4b84e400-23fc-5976-bbb4-f8ee303eed81`

Processing 13405733.pdf...
All specified PDF files have been added.


In [5]:
print("Starting cognify (building the knowledge graph from documents)...")
await cognee.cognify()
print("Cognify process complete.")

Starting cognify (building the knowledge graph from documents)...


INFO:run_tasks(tasks: [Task], data):Coroutine task started: `extract_graph_from_data`INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.13:22:56 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.13:22:56 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.13:22:56 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai13:22:56 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider

Cognify process complete.


In [6]:
await visualize_graph()

# The file is saved in your current folder
home_dir = os.path.expanduser("~")
html_file = os.path.join(home_dir, "graph_visualization.html")
display(html_file)

INFO:cognee.modules.visualization.cognee_network_visualization:Graph visualization saved as /Users/handekafkas/graph_visualization.htmlINFO:root:The HTML file has been stored on your home directory! Navigate there with cd ~

'/Users/handekafkas/graph_visualization.html'

In [64]:
query_1 = (
    "Who managed more than three employees and also had SaaS (or ‘cloud’) responsibilities?"
)

print(f"------ Base RAG with query: '{query_1}' ------")

search_results = await cognee.search(
    query_type=SearchType.COMPLETION,
    query_text=query_1
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")

print(f"------ cognee results with query: '{query_1}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_COMPLETION,
    query_text=query_1
)

print(f"Search results:\n {result_text}\n")



------ Base RAG with query: 'Who managed more than three employees and also had SaaS (or ‘cloud’) responsibilities?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:49:24 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The Information Technology Manager managed a four-person local IT team and had ownership of SaaS customers in North America, Canada, and Australia.

------ cognee results with query: 'Who managed more than three employees and also had SaaS (or ‘cloud’) responsibilities?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:49:33 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
 The Information Technology Manager managed a four-person local IT team and had ownership of SaaS customers in North America, Canada, and Australia.



In [13]:
query_2 = (
    "Which applicants hold (or have trained for) an MCSA certification and also have experience with Active Directory?"
)

print(f"------ Base RAG with query: '{query_2}' ------")

search_results = await cognee.search(
    query_type=SearchType.COMPLETION,
    query_text=query_2
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")


print(f"------ cognee results with query: '{query_2}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_COMPLETION,
    query_text=query_2
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")



------ Base RAG with query: 'Which applicants hold (or have trained for) an MCSA certification and also have experience with Active Directory?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.13:41:50 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The applicants who hold (or have trained for) an MCSA certification and also have experience with Active Directory are qualified professionals listed under the qualifications section. Specifically, they have completed an MCSA certification class and have mentioned experience with Active Directory in their skill set.

------ cognee results with query: 'Which applicants hold (or have trained for) an MCSA certification and also have experience with Active Directory?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.13:41:57 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The applicants who hold (or have trained for) an MCSA certification and have experience with Active Directory include the manager of IT and Building Automation Systems who is certified as a Microsoft Certified Systems Administrator (MCSA), and the IT Management professional who has completed an MCSA certification class and also has experience with Active Directory.



In [71]:
query_3 = (
    "Who has a Master’s degree in IT (or a related field) and also performed network engineering or network administration tasks?"
)

print(f"------ Base RAG with query: '{query_3}' ------")

search_results = await cognee.search(
    query_type=SearchType.COMPLETION,
    query_text=query_3
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")

print(f"------ cognee results with query: '{query_3}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_COMPLETION,
    query_text=query_3
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")



------ Base RAG with query: 'Who has a Master’s degree in IT (or a related field) and also performed network engineering or network administration tasks?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.17:02:16 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The individual has a Master of Science in Computer & Information Science, which is a related field to IT. They have performed network engineering tasks, including managing network infrastructure and server environments, troubleshooting networks, and providing technical support.

------ cognee results with query: 'Who has a Master’s degree in IT (or a related field) and also performed network engineering or network administration tasks?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.17:02:20 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The node that has a Master’s degree in IT (or a related field) and performed network engineering or network administration tasks is the **it manager**.



In [50]:
query_4 = (
    "Who are the candidates who have a degree in business administration and management?"
)

print(f"------ Base RAG with query: '{query_4}' ------")

search_results = await cognee.search(
    query_type=SearchType.COMPLETION,
    query_text=query_4
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")


print(f"------ cognee results with query: '{query_4}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_COMPLETION,
    query_text=query_4
)

print(f"Search results:\n {result_text}\n")


print(f"------ Graph Summary results with query: '{query_4}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_SUMMARY_COMPLETION,
    query_text=query_4
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")

------ Base RAG with query: 'Who are the candidates who have a degree in business administration and management?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:13:00 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The candidates who have a degree in business administration and management are not specifically mentioned in the provided context. The closest relevant degree listed is a Master of Science in Business Information Technology.

------ cognee results with query: 'Who are the candidates who have a degree in business administration and management?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:13:04 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
 The candidates who have a degree in business administration and management are not specifically mentioned in the provided context. The closest relevant degree listed is a Master of Science in Business Information Technology.

------ Graph Summary results with query: 'Who are the candidates who have a degree in business administration and management?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:13:12 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:13:15 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The candidate with a degree in business administration and management is the "DIRECTOR OF INFORMATION TECHNOLOGY."



In [59]:
query_5 = (
    "Who have been to project management institute or holds a PMP certification?"    
)

print(f"------ Base RAG with query: '{query_5}' ------")

search_results = await cognee.search(
    query_type=SearchType.COMPLETION,
    query_text=query_5
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")


print(f"------ cognee results with query: '{query_5}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_COMPLETION,
    query_text=query_5
)

print(f"Search results:\n {result_text}\n")


print(f"------ Graph Summary results with query: '{query_5}' ------")

search_results = await cognee.search(
    query_type=SearchType.GRAPH_SUMMARY_COMPLETION,
    query_text=query_5
)

print("Search results:")
for idx, result_text in enumerate(search_results, 1):
    print(f"{idx}. {result_text}\n")

------ Base RAG with query: 'Who have been to project management institute or holds a PMP certification?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:24:22 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The individual is a current member of the Project Management Institute (PMI) and holds a Certified Associate in Project Management (CAPM) certification.

------ cognee results with query: 'Who have been to project management institute or holds a PMP certification?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:24:26 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
 The individual is a current member of the Project Management Institute (PMI) and holds a Certified Associate in Project Management (CAPM) certification.

------ Graph Summary results with query: 'Who have been to project management institute or holds a PMP certification?' ------


INFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:24:28 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:cognee.infrastructure.llm.utils:Model not found in LiteLLM's model_cost.16:24:30 - LiteLLM:INFO: utils.py:2974 - 
LiteLLM completion() model= gpt-4o-mini; provider = openaiINFO:LiteLLM:
LiteLLM completion() model= gpt-4o-mini; provider = openai

Search results:
1. The Manager in Information Technology and Building Automation Systems holds a Project Management Professional (PMP) certification. Additionally, the Director of Information Technology is associated with the Project Management Institute (PMI).



# Test set

| Query    | Reference | Doc |
| -------- | ------- | -------|
| query_1  | INFORMATION TECHNOLOGY MANAGER | 10553553.pdf|
| query_2 | MANAGER - INFORMATION TECHNOLOGY AND BUILDING AUTOMATION SYSTEMS, IT MANAGEMENT | 11584809.pdf & 10641230.pdf |
| query_3    | INFORMATION TECHNOLOGY MANAGER, MANAGER - INFORMATION TECHNOLOGY AND BUILDING AUTOMATION SYSTEMS | 10553553.pdf & 11584809.pdf |
| query_4    | DIRECTOR OF INFORMATION TECHNOLOGY | 13405733.pdf |
| query_5    | MANAGER - INFORMATION TECHNOLOGY AND BUILDING AUTOMATION SYSTEMS, DIRECTOR OF INFORMATION TECHNOLOGY | 11584809.pdf & 13405733.pdf |
